### PR TITLE
Simplify SolutionDescriptor.getMovableEntityCount()

### PR DESCRIPTION
**Pull request by yurloc**
It's unnecessary to count all movable entities when it's enough to find any and quit early.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

We do "simple" maven builds, they are just basically maven commands, but just because we have multiple repositories related between them and one change could affect several of those projects by multiple pull requests, we use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.

[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is not only a github-action tool but a CLI one, so in case you posted multiple pull requests related with this change you can easily reproduce the same build by executing it locally. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>specific pull request build</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] tests</b>
* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>specific Quarkus LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] LTS</b>
* for a <b>Native check</b> please add comment: <b>Jenkins run native</b>
* for a <b>specific Native LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] native</b>
</details>

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java [hunk 1]
```diff
@@ -931,12 +931,9 @@ public int getEntityCount(Solution_ solution) {
...
      * @return {@code >= 0}
      */
-    public int getMovableEntityCount(ScoreDirector<Solution_> scoreDirector) {
+    public boolean hasMovableEntities(ScoreDirector<Solution_> scoreDirector) {
         return extractAllEntitiesStream(scoreDirector.getWorkingSolution())
-                .mapToInt(entity -> findEntityDescriptorOrFail(entity.getClass()).isMovable(scoreDirector, entity)
-                        ? 1
-                        : 0)
-                .sum();
+                .anyMatch(entity -> findEntityDescriptorOrFail(entity.getClass()).isMovable(scoreDirector, entity));
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:936]: This is a beneficial change on its own. But looking at `findEntityDescriptorOrFail(...)`, it has a performance issue of its own.

It does try to cache the results of the lookup. Unfortunately, when nothing is found, it doesn't store anything in the cache - so in that case (= no entity descriptor found) the lookups will be happening over and over again.

If we're touching this code, I'd vote to address this issue too.

**Review by triceo — APPROVED** [review]
I like this, but I made a suggestion inside.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java (hunk 1 above)
Comment (yurloc) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:933]: @triceo Done. Please take a look.

**Comment by yurloc** [comment]
> I will leave it to him to have the final word. That said, I'd definitely keep the `null` return type for `findEntityDescriptor(...)`, as it would be inconsistent for the API to use `Optional` in this one place and not anywhere else.

I will gladly make the rest of the API consistent by using Optional anywhere where `null` could be returned :slightly_smiling_face:

**Comment by yurloc** [comment]
If there is not a clear consensus on how to address the memoized entity descriptor map optimization, I propose to merge the PR in its original form (the first commit only) and handle the rest separately.

**Comment by triceo** [comment]
> I will gladly make the rest of the API consistent by using Optional anywhere where `null` could be returned slightly_smiling_face

Preaching to the choir here. :-) In this case, I am not the one who needs to be convinced.

> If there is not a clear consensus on how to address the memoized entity descriptor map optimization, I propose to merge the PR in its original form (the first commit only) and handle the rest separately.

Yeah, let's kick the can further down the road.

**Pull request merged**